### ISSUE TITLE
Pass-through ringback in #dial

### ISSUE THREAD

**Issue by dwilkie** [adhearsion/adhearsion#298]
See also https://groups.google.com/forum/?fromgroups=#!topic/adhearsion/ppvcCZKcNU0

When using the dial command in Adhearsion I cannot hear any ringback signal when bridging to a new outgoing-leg. Also when leg-B starts ringing it does not stop when leg-A hangs up.

Adhearsion version git sha: c72670138da1
Punchblock version git sha: d2baea8609b6

Here is the Adhearsion trace where it tries to originate the call:

```
 [2013-05-01 11:04:26] TRACE RubyFS::Stream: [SEND] bgapi originate {return_ring_ready=true,origination_uuid=c7f279eb-eec8-458a-b6e4-89ee1018567f,origination_caller_id_number='+14152345678,originate_timeout=30}sofia/gateway/didlogic/85512345678 &park()
```

And here's some of the trace seen afterwards: https://gist.github.com/dwilkie/5494826

I was thinking that this may have something to do with `return_ring_ready=true` then parking the call. Any thoughts?

**Comment by colinmorelli** [comment]
@benlangfeld Is this something that you anticipate going into 3.0? I need this functionality as well. I was reading the mailing list post and saw that this can be implemented by optimizing to a direct bridge in the event a confirmation controller isn't needed. Is that optimization still valid?

My biggest problem here is actually not even the ringback (I can simulate that with a `play`), but because Adhersion doesn't bridge the calls until they enter an answer state, I sometimes miss the first second or two of audio. While this isn't usually a problem for human -> human calls, if the remote end is an IVR that starts playing audio immediately it can be problematic.

**Comment by benlangfeld** [comment]
> Is this something that you anticipate going into 3.0?

No. It's had no movement for a long time and it doesn't look like anyone's stepping up to change that.

> I was reading the mailing list post and saw that this can be implemented by optimizing to a direct bridge in the event a confirmation controller isn't needed. Is that optimization still valid?

I believe so, but it's a while since I've look at the details of this.

**Comment by colinmorelli** [comment]
Thanks for the quick reply. Are you aware of any workarounds for the missing second or so of audio? Is the current recommendation to use `play` in order to simulate a ringback?

I can look into adding this into adhearsion. I'd have to learn my way around the code (just started using it today), but it's seems like it's doable.

**Comment by benlangfeld** [comment]
> Are you aware of any workarounds for the missing second or so of audio?

**Comment by benlangfeld** [comment]
> Are you aware of any workarounds for the missing second or so of audio?

No, this really is the "workaround".

> Is the current recommendation to use `play` in order to simulate a ringback?

There is a `ringback` option to `#dial`.

**Comment by chewi** [comment]
The good news is that this just works under FreeSWITCH's mod_rayo. In my case, the A-leg is already answered but I don't think that matters. I did have to reverse `call2.join call1` though otherwise it complained that the A-leg was not ready to join.

```ruby
def run
  answer
  call2 = Adhearsion::OutboundCall.new

  call2.register_event_handler Adhearsion::Event::Ringing do
    call.join call2
  end

  call2.dial(somewhere)
  call2.wait_for_end
end
```

**Comment by chewi** [comment]
I have since found that the above sometimes complains that the B-leg is not ready to join. I think there is a race condition here. A `sleep 1` is probably sufficient but to avoid unnecessary delay while also giving it a greater chance, I did this.

```ruby
call2.register_event_handler Adhearsion::Event::Ringing do
  tries = 0

  begin
    sleep 0.1
    tries += 1
    call.join call2
  rescue Adhearsion::ProtocolError
    retry if tries < 50 && $!.name == :unexpected_request
    raise
  end
end
```
